crates/bevy_reflect/src/lib.rs

```diff
@@ -1045,4 +1047,56 @@ bevy_reflect::tests::should_reflect_debug::Test {
             assert_eq!(v, vec3(4.0, 2.0, 1.0));
         }
     }
+
+    #[test]
+    fn type_data_specialization() {
+        #[derive(Reflect, Debug)]
+        #[reflect(MyTrait)]
+        struct HasGenericTypeData;
+
+        #[derive(Reflect, Debug)]
+        #[reflect(MyTrait)]
+        struct HasSpecializedTypeData;
+
+        #[derive(Clone, Debug, PartialEq, Eq)]
+        enum ReflectMyTrait {
+            Generic,
+            Specialized,
+        }
+
+        impl<T> FromType<T> for ReflectMyTrait {
+            fn from_type() -> Self {
+                Self::Generic
+            }
+        }
+
+        impl SpecializedFromType<HasSpecializedTypeData> for ReflectMyTrait {
+            fn specialized_from_type() -> Self {
+                Self::Specialized
+            }
+        }
+
+        let type_registry = {
+            let mut type_registry = TypeRegistry::new();
+            type_registry.register::<HasGenericTypeData>();
+            type_registry.register::<HasSpecializedTypeData>();
+            type_registry
+        };
+
+        assert_eq!(
+            *type_registry
```

Comment:
Is the dereference here (and assertion below) necessary?

Reply:
Nope, removed it and made the right side a reference.